DOCS-5309 HAProxy Integration Edits (#15227)
DataDog/integrations-core: haproxy/README.md

## Setup

This integration can collect metrics from a Prometheus endpoint (recommended) or from a socket-based integration through the stats endpoint (deprecated). Using the Prometheus endpoint requires HAProxy version 2 (enterprise version 1.9rc1) or later.

> When using the Prometheus endpoint, starting with version 4.0.0, this OpenMetrics-based integration has a *latest* mode (`use_openmetrics`: true) and a *legacy* mode (`use_openmetrics`: false and `use_prometheus`: true). To get all the most up-to-date features, Datadog recommends enabling the *latest* mode. For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].

Author: ```suggestion
When using the Prometheus endpoint, starting with version 4.0.0, this OpenMetrics-based integration has a latest mode (`use_openmetrics`: true) and a legacy mode (`use_openmetrics`: false and `use_prometheus`: true). To get all the most up-to-date features, Datadog recommends enabling the latest mode. For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].
```

To use the socket-based integration, set both `use_openmetrics` and `use_prometheus` to false and follow the [corresponding instructions](#using-the-stats-endpoint) on the Configuration section.

> The `use_openmetrics` option uses [OpenMetrics V2][26] for metric collection, which requires Agent v7.35 or later, or for you to [enable Python 3][27] in Agent v6.35 or later. For hosts that are unable to use Python 3 or are on Agent v7.34 or earlier, use the OpenMetrics V1 implementation or the [socket-based legacy integration](#using-the-stats-endpoint). 

Author: ```suggestion
The `use_openmetrics` option uses the latest mode of [OpenMetrics][26], which requires Agent v7.35 or later, or for you to [enable Python 3][27] in Agent v6.35 or later for metric collection. For hosts that are unable to use Python 3 or are on Agent v7.34 or earlier, use the legacy mode of OpenMetrics or the [socket-based legacy integration](#using-the-stats-endpoint). 
```

## Setup

### Prerequisites

> This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version (enterprise version 1.9rc1 or later) to enable the Prometheus endpoint on HAProxy. If you are using the legacy version, use the [HAProxy Prometheus exporter][3], or set up the [socket-based integration](#using-the-stats-endpoint). For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].

Author: ```suggestion
This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version (enterprise version 1.9rc1 or later) to enable the Prometheus endpoint on HAProxy. If you are using the legacy version, use the [HAProxy Prometheus exporter][3], or set up the [socket-based integration](#using-the-stats-endpoint). For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].
```

> This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version (enterprise version 1.9rc1 or later) to enable the Prometheus endpoint on HAProxy. If you are using the legacy version, use the [HAProxy Prometheus exporter][3], or set up the [socket-based integration](#using-the-stats-endpoint). For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].

Reviewer: This is confusing enough that #15449 was recently opened trying to clarify this. Perhaps we can incorporate the idea behind that PR into this one?

> ### Prerequisites
> 
> This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version (enterprise version 1.9rc1 or later) to enable the Prometheus endpoint on HAProxy. If you are using the legacy version, use the [HAProxy Prometheus exporter][3], or set up the [socket-based integration](#using-the-stats-endpoint). For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].

Author: ```suggestion
### Prerequisites

This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version (enterprise version 1.9rc1 or later) to enable the Prometheus endpoint on HAProxy. If you are using the legacy version, use the [HAProxy Prometheus exporter][3], or set up the [socket-based integration](#using-the-stats-endpoint). For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].
```

Reviewer: My take:

```suggestion
This integration can collect metrics from a Prometheus endpoint via (recommended) or via a socket-based integration through the stats endpoint (deprecated). Using the Prometheus endpoint requires HAProxy version 2 (enterprise version 1.9rc1) or later.

When using the Prometheus endpoint, starting with version 4.0.0, this OpenMetrics-based integration has a *latest* mode (`use_openmetrics`: true) and a *legacy* mode (`use_openmetrics`: false and `use_prometheus`: true). To get all the most up-to-date features, Datadog recommends enabling the *latest* mode. For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].

To use the socket-based integration, set both `use_openmetrics` and `use_prometheus` to false and follow the [corresponding instructions](#using-the-stats-endpoint) on the Configuration section.
```

I know it's confusing because of the two layers of legacy that we have on this integration. Hopefully this suggestion addresses that, and helps somewhat with what #15449 addresses. Some of the information becomes redundant with things that are also stated in the configuration section, and I'm not sure how to address that, but I feel that the global view of the "first layer of legacy" needs to be explained this time before getting into the OpenMetrics legacy vs latest stuff.

> This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version (enterprise version 1.9rc1 or later) to enable the Prometheus endpoint on HAProxy. If you are using the legacy version, use the [HAProxy Prometheus exporter][3], or set up the [socket-based integration](#using-the-stats-endpoint). For more information, see the [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].

Reviewer 2: ```suggestion
This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version (enterprise version 1.9rc1 or later) to enable the Prometheus endpoint on HAProxy. If you are using the legacy version, use the [HAProxy Prometheus exporter][3], or set up the [socket-based integration](#using-the-stats-endpoint). For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][29].
```

The `use_openmetrics` option uses [OpenMetrics V2][26] for metric collection, which requires Agent v7.35 or later, or for you to [enable Python 3][27] in Agent v6.35 or later. For hosts that are unable to use Python 3 or are on Agent v7.34 or earlier, use the OpenMetrics V1 implementation or the [socket-based legacy integration](#using-the-stats-endpoint). 

> Metrics marked as `[OpenMetrics V1]` or `[OpenMetrics V2]` are only available using the corresponding version of the HAProxy integration. Metrics marked as `[OpenMetrics V1 and V2]` are collected by both versions.

Author: ```suggestion
To use Prometheus/OpenMetrics v1 instead of OpenMetrics v2, change the `use_openmetrics` option to `use_prometheus`, and change the `openmetrics_endpoint` option to `prometheus_url`. For more information, see the [Prometheus and OpenMetrics metrics collection from a host documentation][29]. Metrics marked as `[OpenMetrics V1]` or `[OpenMetrics V2]` are only available using the corresponding version of the HAProxy integration. Metrics marked as `[OpenMetrics V1 and V2]` are collected by both versions.
```

Reviewer: I would leave this as it was (just the reference to "tags" in metric descriptions), and move the additional information to the configuration section (somewhere in [here](https://github.com/DataDog/integrations-core/blob/0fbfb5a381b96539c2143da8464d25a5985a812d/haproxy/README.md?plain=1#L29)).

#### Using the stats endpoint

> <div class="alert alert-danger">This configuration strategy is provided as a reference for legacy users. If you are setting up the integration for the first time, consider using the Prometheus-based strategy described in the previous section.</div>

Reviewer: Heads up that anything inside HTML tags gets stripped when displaying the contents of the readme in-app, so this change will cause this paragraph to not show up in there.

Author: ```suggestion
This configuration strategy is provided as a reference for legacy users. If you are setting up the integration for the first time, consider using the Prometheus-based strategy described in the previous section.
```

[25]: https://github.com/DataDog/integrations-core/blob/7.34.x/haproxy/datadog_checks/haproxy/data/conf.yaml.example
[26]: https://datadoghq.dev/integrations-core/base/openmetrics/
[27]: https://docs.datadoghq.com/agent/guide/agent-v6-python-3/?tab=helm#use-python-3-with-datadog-agent-v6
[28]: https://github.com/DataDog/integrations-core/blob/0e34b3309cc1371095762bfcaf121b0b45a4e263/haproxy/datadog_checks/haproxy/data/conf.yaml.example#L631

> [29]: https://docs.datadoghq.com/integrations/guide/versions-for-openmetrics-based-integrations

Author: ```suggestion
[29]: https://docs.datadoghq.com/integrations/guide/versions-for-openmetrics-based-integrations
[30]: https://docs.datadoghq.com/integrations/guide/prometheus-host-collection/
```

[25]: https://github.com/DataDog/integrations-core/blob/7.34.x/haproxy/datadog_checks/haproxy/data/conf.yaml.example
[26]: https://datadoghq.dev/integrations-core/base/openmetrics/
[27]: https://docs.datadoghq.com/agent/guide/agent-v6-python-3/?tab=helm#use-python-3-with-datadog-agent-v6
[28]: https://github.com/DataDog/integrations-core/blob/0e34b3309cc1371095762bfcaf121b0b45a4e263/haproxy/datadog_checks/haproxy/data/conf.yaml.example#L631

> [29]: https://docs.datadoghq.com/integrations/guide/versions-for-openmetrics-based-integrations

Author: ```suggestion
[29]: https://docs.datadoghq.com/integrations/guide/versions-for-openmetrics-based-integrations
[30]: https://docs.datadoghq.com/integrations/guide/prometheus-host-collection/
```
